### ISSUE TITLE
std::pair<U, V> bindings are broken

### ISSUE THREAD

**Issue by robinchrist** [taocpp/json#63]
I'll keep it as short as possible:

Works:
`obj.at_key("firstPickpointKey").template as<rigging::PickpointKey>(),`

Does not work:
`obj.at_key("secondPickpointKey").template as<std::pair<rigging::PickpointKey, float>>()`

with error
```
../../../../../projectname/include/taocpp/json/include/tao/json/binding/internal/../../basic_value.hpp:1273:12: error: no matching constructor for initialization of 'std::pair<CVO::array::rigging::PickpointKey, float>'
         T v;
           ^
../../../../../projectname/include/taocpp/json/include/tao/json/contrib/../traits.hpp:887:28: note: in instantiation of function template specialization 'tao::json::basic_value<traits>::as<std::pair<CVO::array::rigging::PickpointKey, float> >' requested here
         return v.template as< T >();
                           ^
../../../../../projectname/include/taocpp/json/include/tao/json/binding/internal/../../basic_value.hpp:1267:30: note: in instantiation of function template specialization 'tao::json::traits<std::optional<std::pair<CVO::array::rigging::PickpointKey, float> >, void>::as<traits>' requested here
         return Traits< T >::as( *this );
```

I've patched the pair traits temporarily (manually instead of using the array binding traits)

**Comment by ColinH** [comment]
My feeling is that the problem is a combination of `rigging::PickpointKey` not being default-constructible and `binding::array` only offering `to()`, which requires the target type to be default-constructed. The `as()` in `basic_value` falls back to `to()` when it can't find an `as()` in the traits, but default-constructs the return value. The solution would be to have `binding::array` also offer `as()` in case the target type can be constructed with the converted elements of the array. I'll see whether that's possible.

**Comment by ColinH** [comment]
How exactly did you patch the pair traits to construct the `rigging::PickpointKey`, and how do you construct it in the traits for `rigging::PickpointKey` that makes `obj.at_key("firstPickpointKey").template as<rigging::PickpointKey>()` work?

**Comment by robinchrist** [comment]
For the first point, see https://github.com/robinchrist/json/commit/463f9db1ca9086e65b1841fc362dac490c701ac9

That's most probably not the optimal solution, however in my case, it makes things work until you have figured out a better solution...

Regarding the second point:

```
template<> struct tao::json::traits<rigging::PickpointKey> {

    template<template<typename...> class Traits>
    static rigging::PickpointKey as(const tao::json::basic_value<Traits>& v ) {
        const auto& object = v.get_object();

        return rigging::PickpointKey {
            object.at_key("firstKey").template as<unsigned int>(),
            object.at_key("secondKey").template as<unsigned int>()
        };
   }
};
```

**Comment by ColinH** [comment]
Ok thanks, that confirms my understanding of the situation. Now towards adding an `as()`-function to the array-binding-traits, there's a snag.

The target type is not passed as template parameter to the traits functions, it is assumed that each traits (specialisation) knows for which type it is responsible, and therefore a potential `binding::array<>::as()`-function does not have the type it is supposed to return available (unlike `to()` which receives a reference to an already existing object).

@d-frey We need to look into this and decide on whether passing the type to `traits::as()` and `traits::consume()` would be worth the hassle, whether there is another solution, or whether we'll just assume that the types are default constructible in this context.

**Comment by ColinH** [comment]
Does it work without changing the pair-traits with the latest commit?

**Comment by robinchrist** [comment]
Sorry for the delay. I'm gonna try that later and get back to you asap!

**Comment by robinchrist** [comment]
I've merged everything until fb57b210bc46c0b3ab303d28ba4cfb7e93fa6640 and it seems to work, I'm going to do some more testing!

**Comment by robinchrist** [comment]
I have experienced no issues so far. If there are any issues, I'll open another issue.

Thanks a lot, great work!
